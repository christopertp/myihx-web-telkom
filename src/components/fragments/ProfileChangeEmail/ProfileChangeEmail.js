import React, { createContext, useContext, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useParams } from 'react-router-dom';
import WhiteCardPlain from '../../layouts/WhiteCardPlain';
import ChangeEmailForm from '../../forms/ProfileChangeEM';
import RecoveryCodeForm from '../../forms/ProfileRecoveryCode';

const Context = createContext({});

export default function ProfileChangeEmail() {
  const { subpage } = useParams();
  const { email, resetMessage } = useSelector(s => s.profile);
  const dispatch = useDispatch();
  const history = useHistory();
  const [ emailUser, setEmailUser ] = useState(email);

  const contextValue = {
    emailUser,
    setEmailUser,
  };

  useEffect(() => {
    subpage && history.replace('/profile/change-email');
  }, []);

  useEffect(() => {
    dispatch(resetMessage());
  }, [subpage]);

  return (
    <Context.Provider value={contextValue}>
      {(subpage === 'verify') ? <Verify /> : <New />}
    </Context.Provider>
  );
}

export function New() {
  const { setEmailUser } = useContext(Context);
  const { fetchChangeEmail } = useSelector (s => s.profile);
  const dispatch = useDispatch();
  const onSubmit = v => {
    setEmailUser(v.email);
    dispatch(fetchChangeEmail(v.email));
  };

  return (
    <WhiteCardPlain back="/profile" help icon="email" size="medium" title="Change Email">
      <ChangeEmailForm onSubmit={onSubmit} />
    </WhiteCardPlain>
  );
}

export function Verify() {
  const { emailUser } = useContext(Context);
  const { fetchChangeEmail, fetchVerifyEmail } = useSelector(s => s.profile );
  const subtitle = 'Please enter verification code sent to your previous email';
  const dispatch = useDispatch();
  const onResend = () => dispatch(fetchChangeEmail(emailUser));
  const onSubmit = v => dispatch(fetchVerifyEmail(v.recoveryCode));

  return (
    <WhiteCardPlain back="/profile/change-email" help icon="email" size="medium"
      subtitle={subtitle} title="Verify Email">
      <RecoveryCodeForm onResend={onResend} onSubmit={onSubmit} />
    </WhiteCardPlain>
  );
}
