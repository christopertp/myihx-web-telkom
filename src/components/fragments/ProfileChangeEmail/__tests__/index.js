import React, { useContext } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import ShallowRenderer from 'react-test-renderer/shallow';
import ProfileChangeEmail, { New, Verify } from '../ProfileChangeEmail';

useSelector.mockImplementation(fn => {
  fn({});
  return {
    email: 'test@telkom.co.id',
    fetchChangeEmail: jest.fn(),
    fetchVerifyEmail: jest.fn(),
    resetMessage: jest.fn(),
  };
});

describe('src/components/fragments/ProfieChangeEmail', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<ProfileChangeEmail />);
    expect(tree).toMatchSnapshot();

    useParams.mockImplementationOnce(() => ({ subpage: 'verify' }));
    const result = ProfileChangeEmail();
    expect(result.props.children.type).toBe(Verify);
  });

  test('New', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<New />);
    expect(tree).toMatchSnapshot();

    const setEmailUser = jest.fn();
    const dispatch = jest.fn();
    useContext.mockImplementationOnce(() => ({ setEmailUser }));
    useDispatch.mockImplementationOnce(() => dispatch);
    const result = New();
    result.props.children.props.onSubmit({ email: 'test2@telkom.co.id' });
    expect(setEmailUser).toHaveBeenCalledWith('test2@telkom.co.id');
    expect(dispatch).toHaveBeenCalled();
  });

  test('Verify', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Verify />);
    expect(tree).toMatchSnapshot();

    const emailUser = jest.fn();
    const dispatch = jest.fn();
    useContext.mockImplementationOnce(() => ({ emailUser }));
    useDispatch.mockImplementationOnce(() => dispatch);
    const result = Verify();
    result.props.children.props.onResend();
    result.props.children.props.onSubmit({ recoveryCode: '123' });
    expect(dispatch).toHaveBeenCalledTimes(2);
  });
});
