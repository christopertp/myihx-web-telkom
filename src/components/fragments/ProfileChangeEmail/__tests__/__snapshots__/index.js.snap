// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/fragments/ProfieChangeEmail New 1`] = `
<WhiteCardPlain
  back="/profile"
  close=""
  help={true}
  icon="email"
  size="medium"
  subtitle=""
  title="Change Email"
>
  <ReduxForm
    destroyOnUnmount={true}
    enableReinitialize={false}
    forceUnregisterOnUnmount={false}
    form="profileChangeEM"
    getFormState={[Function]}
    initialValues={
      Object {
        "email": "",
        "mobile": "",
      }
    }
    keepDirtyOnReinitialize={false}
    onSubmit={[Function]}
    persistentSubmitErrors={false}
    pure={true}
    shouldAsyncValidate={[Function]}
    shouldError={[Function]}
    shouldValidate={[Function]}
    shouldWarn={[Function]}
    submitAsSideEffect={false}
    touchOnBlur={true}
    touchOnChange={false}
    updateUnregisteredFields={false}
    validate={[Function]}
  />
</WhiteCardPlain>
`;

exports[`src/components/fragments/ProfieChangeEmail Verify 1`] = `
<WhiteCardPlain
  back="/profile/change-email"
  close=""
  help={true}
  icon="email"
  size="medium"
  subtitle="Please enter verification code sent to your previous email"
  title="Verify Email"
>
  <ReduxForm
    destroyOnUnmount={true}
    enableReinitialize={false}
    forceUnregisterOnUnmount={false}
    form="profileRecoveryCode"
    getFormState={[Function]}
    initialValues={
      Object {
        "recoveryCode": "",
      }
    }
    keepDirtyOnReinitialize={false}
    onResend={[Function]}
    onSubmit={[Function]}
    persistentSubmitErrors={false}
    pure={true}
    shouldAsyncValidate={[Function]}
    shouldError={[Function]}
    shouldValidate={[Function]}
    shouldWarn={[Function]}
    submitAsSideEffect={false}
    touchOnBlur={true}
    touchOnChange={false}
    updateUnregisteredFields={false}
    validate={[Function]}
  />
</WhiteCardPlain>
`;

exports[`src/components/fragments/ProfieChangeEmail render 1`] = `
<Provider
  value={
    Object {
      "emailUser": "test@telkom.co.id",
      "setEmailUser": [MockFunction],
    }
  }
>
  <New />
</Provider>
`;
